components/TextFields/MDCTextInput.h

```diff
@@ -0,0 +1,191 @@
+ */
+@property(nonatomic, nullable, strong) UIColor *floatingPlaceholderColor UI_APPEARANCE_SELECTOR;
+
+/** The text displayed in the text input. */
+@property(nonatomic, nullable, copy) NSString *text;
+
+/**
+ The text displayed under input rectangle, when in error state, that explains the nature of the
+ error.
+
+ The value of this property controls the state of the text input. When errorText != nil, the text
+ input is in an error state:
+  - The error text appears in a label below the input rectangle's underline with the errorColor as
+  text color.
+  - The input rectangle's underline is colored with the errorColor.
+ */
+@property(nonatomic, nullable, copy, readonly) IBInspectable NSString *errorText;
+
+/**
+ The color used to denote error state in the underline and the errorText's label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, copy) UIColor *errorColor UI_APPEARANCE_SELECTOR;
+
+/**
+ A localized string that represents the value of the errorText label. Use only when the you need
+ to override the default which is the errorText itself.
+ */
+@property(nonatomic, nullable, copy, readonly) NSString *errorAccessibilityText;
+
+/** The delegate for the text input. */
+@property(nonatomic, nullable, weak) IBInspectable id<MDCTextInputLayoutDelegate> layoutDelegate;
+
+/**
+ The presentation style of the text input.
+
+ Default is MDCTextInputPresentationStyleDefault.
+ */
+@property(nonatomic, assign)
+    MDCTextInputPresentationStyle presentationStyle NS_SWIFT_NAME(presentation);
+
+/**
+ The character limit for the text input.
+
+ Displays the character limit text taking into account the MDCTextInputPresentationStyle. A value
+ of 0 removes the limit text.
+
+ Default is 0.
+ */
+@property(nonatomic, assign) IBInspectable NSUInteger characterLimit;
+
+/**
+ The color applied to the character count / character limit label. However, when in error state,
+ it will be colored with the error color.
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong) UIColor *characterCountColor UI_APPEARANCE_SELECTOR;
+
+/**
+ The color applied to the underline. However, when in error state, it will be colored with the
+ error color.
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong) UIColor *underlineColor UI_APPEARANCE_SELECTOR;
+
+/**
+ The view mode for the underline.
+
+ Default is UITextFieldViewModeAlways.
+ */
+@property(nonatomic, assign) UITextFieldViewMode underlineViewMode NS_SWIFT_NAME(underlineMode);
+
+/**
+ Set the values of properties: errorText, errorColor, errorAccessibilityValue.
+
+ The value of errorText controls the state of the text input.
+
+ @param errorText               The errorText to be shown under the input rectangle's underline.
+ @param errorAccessibilityValue Optional override of default errorText accessibility value
+ (errorText.text).
+
+ When errorText != nil, the text input is in an error state:
```

Comment:
This seems like a lot of logic exposed in the comment. It is a bit of code smell that the error color magic happens based off of weather the text is something.
I feel like whoever is calling this should be responsible for just setting the values of the properties themselves. Or we could provide another external object/function that does this magic calculation based off the errorText value.